skmultilearn/model_selection/iterative_stratification.py

```diff
@@ -0,0 +1,207 @@
+                self.per_row_combinations[i].append(y)
+
+        self.all_combinations = [list(x) for x in set(self.all_combinations)]
+
+        self.desired_samples_per_label_per_fold = {
+            i:
+               [len(self.samples_with_combination[i]) * self.percentage_per_fold[j]
+                for j in range(self.n_splits)]
+                for i in self.samples_with_combination
+        }
+
+    def label_tie_break(self, M):
+        if len(M) == 1:
+            return M[0]
+        else:
+            max_val = max(self.desired_samples_per_fold[M])
+            M_prim = np.where(
+                np.array(self.desired_samples_per_fold) == max_val)[0]
+            M_prim = np.array([x for x in M_prim if x in M])
+            return np.random.choice(M_prim, 1)[0]
+
+    def get_most_desired_combination(self):
+        currently_chosen = None
+        currently_best_score = None
+        for combination, evidence in self.samples_with_combination.items():
+            current_score = (len(set(combination)), len(evidence))
+            if current_score[1] == 0:
+                continue
+            if currently_chosen is None:
+                currently_chosen = combination
+                currently_best_score = current_score
+                continue
+
+            if currently_best_score[1] > current_score[1] and currently_best_score[0] < current_score[0]:
+                print(currently_best_score, current_score)
+                currently_chosen = combination
+                currently_best_score = current_score
+
+        if current_score is not None:
+            return currently_chosen
+
+        return None
+
+    def distribute_positive_evidence(self):
+        l = self.get_most_desired_combination()
+        while l is not None:
+            print(l)
+            while len(self.samples_with_combination[l]) > 0:
+                row = self.samples_with_combination[l].pop()
+                print(len(self.samples_with_combination[l]), l, row)
+                if self.rows_used[row]:
+                    continue
+
+                max_val = max(self.desired_samples_per_label_per_fold[l])
+                M = np.where(
+                    np.array(self.desired_samples_per_label_per_fold[l]) == max_val)[0]
+                m = self.label_tie_break(M)
+                self.folds[m].append(row)
+                self.rows_used[row] = True
+                for i in self.per_row_combinations[row]:
+                    if row in self.samples_with_combination[i]:
+                        self.samples_with_combination[i].remove(row)
+                    self.desired_samples_per_label_per_fold[i][m] -= 1
+                self.desired_samples_per_fold[m] -= 1
+
+            l = self.get_most_desired_combination()
+
+    def distribute_negative_evidence(self):
+        self.available_samples = [
+            i for i, v in self.rows_used.items() if not v]
+        self.samples_left = len(self.available_samples)
```

Comment:
Maybe @ChristianSch  can also weigh-in on this. But I find that assigning `self.samples_left` as an attribute is a bit costly. I see the reasoning for it, because we're deducting `-1` per iteration. But maybe we can write it this way:

```python
for i in range(len(self.available_samples)):
    #do something
```

In this case, we're relieve with the burden of decrementing `self.samples_left`, and we don't need to assign an attribute into it.

Reply:
I agree, fixed.